// @vuepress/helper/lib/client/utils/env.js
export const checkIsMobile = (ua) => /\b(?:Android|iPhone)/i.test(ua);
export const checkIsChromeWebView = (ua) => / wv\).+(chrome)\/([\w.]+)/i.test(ua);
export const checkIsSafariMobile = (ua) => /version\/([\w.]+) .*mobile\/\w+ (safari)/i.test(ua);
export const checkIsSafari = (ua) => /version\/([\w.]+) .*(mobile ?safari|safari)/i.test(ua);
export const checkIsiPhone = (ua) => /\((ip(?:hone|od)[\w ]*);/i.test(ua);
export const checkIsiPad = (ua) => [
    /\((ipad);[-\w),; ]+apple/i,
    /applecoremedia\/[\w.]+ \((ipad)/i,
    /\b(ipad)\d\d?,\d\d?[;\]].+ios/i,
].some((item) => item.test(ua));
export const checkIsWindows = (ua) => [
    /microsoft (windows) (vista|xp)/i,
    /(win(?=3|9|n)|win 9x )([nt\d.]+)/i,
    /(windows) nt 6\.2; (arm)/i,
    /(windows (?:phone(?: os)?|mobile))[/ ]?([\d.\w ]*)/i,
    /(windows)[/ ]?([ntce\d. ]+\w)(?!.+xbox)/i,
].some((item) => item.test(ua));
export const checkIsIOS = (ua) => [
    /ip[honead]{2,4}\b(?:.*os ([\w]+) like mac|; opera)/i,
    /cfnetwork\/.+darwin/i,
].some((item) => item.test(ua));
export const checkIsMacOS = (ua) => [/(mac os x) ?([\w. ]*)/i, /(macintosh|mac_powerpc\b)(?!.+haiku)/i].some((item) => item.test(ua));
